common/src/main/java/com/wynntils/wynn/model/CharacterManager.java

```diff
@@ -199,13 +231,34 @@ public static CharacterInfo parseCharacterFromCharacterMenu(ItemStack itemStack,
             }
             ClassType classType = ClassType.fromName(className);
 
-            return new CharacterInfo(classType, classType != null && ClassType.isReskinned(className), level, id);
+            Map<ProfessionInfo.ProfessionType, Integer> levels = new HashMap<>();
+            List<String> professionLore = ItemUtils.getLore(professionInfoItem);
+            for (String line : professionLore) {
+                Matcher matcher = INFO_MENU_PROFESSION_LORE_PATTERN.matcher(line);
+
+                if (matcher.matches()) {
+                    levels.put(
+                            ProfessionInfo.ProfessionType.valueOf(matcher.group(2)),
+                            Integer.parseInt(matcher.group(1)));
+                }
+            }
+
+            return new CharacterInfo(
+                    classType,
+                    classType != null && ClassType.isReskinned(className),
+                    level,
+                    id,
+                    new ProfessionInfo(levels));
         }
 
         @Override
         public String toString() {
-            return "CharacterInfo[classType=" + classType + ", reskinned=" + reskinned + ", level=" + level + ", id="
-                    + id + ']';
+            return "CharacterInfo{" + "classType="
+                    + classType + ", reskinned="
+                    + reskinned + ", level="
+                    + level + ", id="
+                    + id + ", professionInfo="
+                    + professionInfo + '}';
         }
```

Comment:
Qodana has a point. It makes sense to have a proper toString for ProfessionInfo.